pkg/services/scm_provider/bitbucket_server.go

```diff
@@ -0,0 +1,221 @@
+	paged := map[string]interface{}{
+		"limit": 100,
+	}
+	repos := []*Repository{}
+	for {
+		response, err := b.client.DefaultApi.GetRepositoriesWithOptions(b.projectKey, paged)
+		if err != nil {
+			return nil, fmt.Errorf("error listing repositories for %s: %v", b.projectKey, err)
+		}
+		repositories, err := bitbucketv1.GetRepositoriesResponse(response)
+		if err != nil {
+			log.Errorf("error parsing repositories response '%v'", response.Values)
+			return nil, fmt.Errorf("error parsing repositories response %s: %v", b.projectKey, err)
+		}
+		for _, bitbucketRepo := range repositories {
+			var url string
+			switch cloneProtocol {
+			// Default to SSH if unspecified (i.e. if "").
+			case "", "ssh":
+				url = getCloneURLFromLinks(bitbucketRepo.Links.Clone, "ssh")
+			case "https":
+				url = getCloneURLFromLinks(bitbucketRepo.Links.Clone, "http")
+			default:
+				return nil, fmt.Errorf("unknown clone protocol for Bitbucket Server %v", cloneProtocol)
+			}
+
+			org := bitbucketRepo.Project.Key
+			repo := bitbucketRepo.Name
+			// Bitbucket doesn't return the default branch in the repo query, fetch it here
+			branch, err := b.getDefaultBranch(org, repo)
+			if err != nil {
+				return nil, err
+			}
+			if branch == nil {
+				log.Debugf("%s/%s does not have a default branch, skipping", org, repo)
+				continue
+			}
+
+			repos = append(repos, &Repository{
+				Organization: org,
+				Repository:   repo,
+				URL:          url,
+				Branch:       branch.DisplayID,
+				SHA:          branch.LatestCommit,
+				Labels:       []string{}, // Not supported by library
+				RepositoryId: bitbucketRepo.ID,
+			})
+		}
+		hasNextPage, nextPageStart := bitbucketv1.HasNextPage(response)
+		if !hasNextPage {
+			break
+		}
+		paged["start"] = nextPageStart
+	}
+	return repos, nil
+}
+
+func (b *BitbucketServerProvider) RepoHasPath(_ context.Context, repo *Repository, path string) (bool, error) {
+	opts := map[string]interface{}{
+		"limit": 100,
+		"at":    repo.Branch,
+	}
+	// No need to query for all pages here
+	response, err := b.client.DefaultApi.StreamFiles_42(repo.Organization, repo.Repository, path, opts)
+	if response != nil && response.StatusCode == 404 {
+		// The path requested does not exist at the supplied commit.
+		return false, nil
+	}
+	if response != nil && response.StatusCode == 400 {
+		// If the path is a file, the first call will return 400: The path requested is not a directory at the supplied commit.
+		// Simply retry with an API call that works with files and expect a 200 return code
+		opts["type_"] = true // Only request the type, we don't need the content
+		response, err := b.client.DefaultApi.GetContent_0(repo.Organization, repo.Repository, path, opts)
+		if response != nil && response.StatusCode == 404 {
+			// File not found
+			return false, nil
+		}
+		if err != nil {
+			return false, err
+		}
+		// 200 ok
+		return true, nil
+	}
+	if err != nil {
+		return false, err
+	}
+	return true, nil
```

Comment:
Ditto, do we know that no error necessarily means yes-directory? (I'm thinking of redirects, for example, as maybe being something where `nil != nil` but still does not imply success.)

Reply:
Can't think of a case where it would not work, added a test with a redirect